### PR TITLE
Refactor parameters

### LINKED ISSUE #256: can't save Model objects with xarray properties
When saving a model by itself or as part of a results object, pack_attrs or equivalent is never run, so we have yaml.dump of xarray objects happening. These can't be loaded back up.

Possible solutions:
1) rewrite pack_attrs to work on model, etc.
2) write yaml constructor and representer classes for xarray objects

### LINKED ISSUE #353: Can't fit multiple particles
Right now you can't use the convenient `fit` function on `Scatterers` objects. 

This is because `hp.inference.interface.parameterize_scatterer` calls `hp.scattering.scatterer.scatterer._interpret_parameters`, which was changed in #279 to colons in parameter names as special characters to denote dictionaries or xarrays of parameter values. Unfortunately `hp.scattering.scatterer.Scatterers.parameters` also uses colons to denote different particles within a `Scatterers` object. This hadn't been noticed as a problem because `_interpret_parameters` is mostly used in `from_parameters`, and `Scatterers` defines their own.

### PR DESCRIPTION
**Pull request by barkls**
Fixes #353. Fixes #256.

Moved all parameter handling from Scatterer class to Model class. This fixed some problems with tied parameters and model flexibility. The Model now stores scatterer, alpha, optics, and hopefully theory in the form of a "map" that tells it how to reconstruct the object with appropriate parameter substitutions. This PR largely rewrites the model.py file and that is the most important place to look to understand what's going on. Everything else is mostly just adapting to those changes.

There are a few scatterer API changes that I want to mention:
1) scatterer.parameters used to give a flat dictionary, e.g. `{n.real: Uniform(1.58, 1.60), n.imag: 0.001, 'center.0': 10, 'center.1':10, 'center.2':Uniform(0, 100)}`. After this PR it matches the initial arguments to call the scatterer: `{'n': ComplexPrior(Uniform(1.58, 1.60), 0.001), 'center': (10, 10, Uniform(0, 100))}` .
2) `scatterer.from_parameters` used to always give a definite value - it would substitute all priors with their best guess. With this PR it just returns priors without modification.
3) Previously you could call `calc_holo` on a parameterized scatterer and it would similarly substitute priors with their best guess. This PR requires you to define a model object if you want to do that (but you were going to define a model anyway if using a parameterized scatterer).

There's also a big model API change that I'm less sure about and would appreciate some feedback on:
Previously, model methods (lnprior`, lnlike, residuals, lnposterior, forward, and the new scatterer_from_parameters) wanted a dictionary of parameter values to evaluate at, e.g. {'r': 0.5, 'n': 1.59, 'alpha': 0.8}. With the changes in this PR, I found a list of parameter values was more natural, and there was a lot of switching back and forth between them if we wanted to keep dictionaries everywhere. After this PR, the methods expect a list of parameter values in a specific order e.g. [0.5, 1.59, 0.8]. They can still take a dictionary for backwards compatibility but it is deprecated.

Unfortunately both formats have advantages, and this is where I'd like input. The dictionary is more human readable and lets you easily test how different parameter values would affect the hologram, for example. It matches `model.parameters` which is the main user-facing way to see what the model parameters are. A list makes for cleaner code, and a simpler interface for inference algorithms, both within holopy an externally. This corresponds to `model._parameters`, which does all the work but isn't user facing.

### CODE REVIEW

**Comment by briandleahy** [comment]
It looks like some of the tests are failing. Do you want me to take a look at it now, or wait until you've addressed the tests?

**Comment by barkls** [comment]
Failing tests should be trivial to fix. Review now will be helpful since I expect there might be a few rounds of changes. Thanks!

File: holopy/core/holopy_object.py
Comment (briandleahy) [inline holopy/core/holopy_object.py:110]: Thanks for deleting code that wasn't being used!

File: holopy/core/metadata.py
Comment (briandleahy) [inline holopy/core/metadata.py:455]: again, thanks for deleting unused code!

File: holopy/inference/interface.py
Comment (briandleahy) [reply holopy/inference/interface.py:79]: Did you find this because a test failed? If so, it might be a sign that we need to test this better. (Unless if the parameters don't actually need to be a list.)

File: holopy/inference/interface.py
Comment (briandleahy) [inline holopy/inference/interface.py:109]: I'm a little confused as to what this code does, esp. since you said in the PR that
> scatterer.parameters used to give a flat dictionary, e.g. {n.real: Uniform(1.58, 1.60), n.imag: 0.001, 'center.0': 10, 'center.1':10, 'center.2':Uniform(0, 100)}. After this PR it matches the initial arguments to call the scatterer: {'n': ComplexPrior(Uniform(1.58, 1.60), 0.001), 'center': (10, 10, Uniform(0, 100))} .

Can you let me know what I'm missing, either in the comments or when we meet?

File: holopy/inference/interface.py
Comment (briandleahy) [inline holopy/inference/interface.py:132]: Same as above, here. (I'm a little confused as to what this is doing, since it seems to assume keys have a colon in them, when my impression of the new version was that keys had no colon.

It might be a little clearer if you split out the code that gets the guess value?? Or maybe I'm missing something completely.

File: holopy/inference/nmpfit.py
Comment (briandleahy) [inline holopy/inference/nmpfit.py:167]: I know this is from before your commit, but it would be good if we could change the single-character variable name to something more descriptive.

File: holopy/inference/tests/common.py
Comment (briandleahy) [inline holopy/inference/tests/common.py:41]: Just a comment so we can talk about this in our next meeting -- if the lnposterior for SimpleModel is so simple, why is the sampling so slow?

**Comment by barkls** [comment]
tests should pass now. Sorry to push while you're mid-review!

File: holopy/inference/tests/test_interface.py
Comment (briandleahy) [inline holopy/inference/tests/test_interface.py:199]: Does `parameterize_scatterer(Sphere(center=[0, 0, 0]), ['center'])` return all 3 priors for x, y, and z, or just x? If so, we could do
```python
for i, xyz in enumerate('xyz'):
    expected = prior.Uniform(-np.inf, np.inf, 0, xyz)
    self.assertEqual(scatterer.center[i], expected)
```

File: holopy/inference/interface.py
Comment (briandleahy) [reply holopy/inference/interface.py:109]: Oh... is this because of `Spheres` vs `Sphere`?

File: holopy/scattering/interface.py
Comment (briandleahy) [inline holopy/scattering/interface.py:198]: ZOMG, did you eliminate the need for this ugly massaging of scaling with color channels??!!!?! If so, :100:

File: holopy/scattering/scatterer/scatterer.py
Comment (briandleahy) [inline holopy/scattering/scatterer/scatterer.py:124]: Why the deepcopy? What gets modified in place if you do a regular copy?

File: holopy/scattering/theory/scatteringtheory.py
Comment (briandleahy) [inline holopy/scattering/theory/scatteringtheory.py:282]: Since this is just for multi-wavelength, can we rename this to `_select_multiwavelength_scatterer` or something?

Also, what was happening before -- was there a bug that you caught or is this a new need?

File: holopy/inference/model.py
Comment (briandleahy) [inline holopy/inference/model.py:56]: Are the `map_entry`s seen by the user? If not, it might be worth making a small class / namedtuple that captures this. You could then subsume some of this more complicated logic into that class, which might reduce the amount of edge cases that can happen when an incorrect argument is passed to as `map_entry`.

Comment (briandleahy) [reply holopy/inference/model.py:56]: The advantage would be that errors would be raised at the correct point. For instance, if you call `map_entry.read_map(parameter_values)`, an error will be raised if `map_entry` is a list, whereas currently things will propagate down the chain and make it harder to debug.

Comment (briandleahy) [reply holopy/inference/model.py:56]: It would also have an advantage of making a `map` a concrete object, which would allow for an easier understanding of how the code works. Right now it seems that the mapping functionality is diffused across multiple regions of the file; making it a helper class would concentrate it in one place.

**Review by briandleahy — COMMENTED** [review]
Hey @barkls , 

Thanks for doing this. This was clearly a lot of work. 

In general I think things look good. I'm still a bit confused about how the map actually works. Can you explain it to me when we meet next?

I think the three API changes that you bulleted are good. Regarding the dictionary vs the list -- I think this is OK, as long as there is a clear correct order for the parameters to get passed in. One advantage of a dictionary is that the order does not matter. An alternative is to take a dict as inputs and cast to a list right away for the actual inner workings, although I'm not sure if that would be viable.  It would help if you showed me in the code where this is, when we talk about it.

Regarding changes -- I left a few general comments throughout; feel free to take them or leave them. The pep8 bot also made some comments; it would be nice if you could address some of them. (I appreciate you waiting until I took a look at this first though, thanks!)

When we meet next, we can chat about this and make sure I'm understanding things correctly. Then I'll approve and merge. 

Thanks again @barkls !

File: holopy/scattering/interface.py
Comment (briandleahy) [reply holopy/scattering/interface.py:198]: It is now taken care of by `scatteringtheory.py` and other places. :100:

File: holopy/inference/interface.py
Comment (barkls) [reply holopy/inference/interface.py:79]: Turned out the list casting wasn't actually necessary here, so I've reverted the change.

File: holopy/inference/interface.py
Comment (barkls) [reply holopy/inference/interface.py:109]: Yeah this is to handle the case of Scatterers objects, which might have '2:y' as a user-specified fitting parameter that needs to replaced with '2:center'[1] since that's how it would appear in Spheres.parameters.

File: holopy/inference/interface.py
Comment (barkls) [reply holopy/inference/interface.py:132]: Colons are still used by composite scatterers to denote their individual components.

Automated detection of fitting parameters from scatterers got complicated when scatterers forgot how to interpret themselves. Specifically, the combination of Spheres objects and trying to fit one or more 'center' coordinates ended up producing a bunch of cases that need to be handled. This is why I had suggested not doing the automatic detection for composite scatterers. 

Note that all of this parameter handling in interface.py is just to come up with a Scatterer object containing prior objects for the case where a user calls hp.fit(data, scatterer). More advanced users will define their own Scatterer with prior objects and a Model.

File: holopy/inference/nmpfit.py
Comment (barkls) [reply holopy/inference/nmpfit.py:167]: renamed to parameters

File: holopy/inference/tests/common.py
Comment (barkls) [reply holopy/inference/tests/common.py:41]: see email thread...

File: holopy/scattering/theory/scatteringtheory.py
Comment (barkls) [reply holopy/scattering/theory/scatteringtheory.py:282]: renamed to `select_scatterer_by_illumination`. In theory it should work for e.g. cross-polarized light sources as well as different wavelengths. 

This was previously handled by the Scatterer itself, but has been moved here as part of simplifying the Scatterer class.

File: holopy/inference/model.py
Comment (barkls) [reply holopy/inference/model.py:56]: I tried making a class to handle this but it didn't help as much as I hoped due to the following:
1) Some of the map interpretation code couldn't be  easily incorporated into the new class since it was needed to read raw maps (not part of an object) during Model serialization/unserialization.
2) There ended up being a lot of calling back and forth between the model and the new object and I needed to add a whole bunch of new methods/properties to the model class just to retrieve things from the new object while maintaining existing namespaces.
3) It became more convoluted to add a new type of Model, since the forward methods would need to interact with the new class.

All those difficulties aside, it did nicely store all the data currently associated with the Model class, and the Model became a series of wrapper methods specifically for doing inference. This matches nicely with something I remember reading in Clean Code about having classes to store data and classes to do calculations, but ideally not both in the same object.

I agree this would be a good change to make, but I couldn't figure out an elegant way to do it, so I abandoned my efforts. I'm happy to revisit it if anyone thinks the current state of model.py is too messy and/or have suggestions about resolving the above issues.

**Comment by barkls** [comment]
I've applied @briandleahy suggestions here, and made a clearer distinction among Model methods between those that are user-facing (cast parameters dict to list if necessary), and those that are used internally (only accept parameters list).

I didn't end up making a separate class to hold the model parameters and map - it ended up being too tightly integrated into the existing Model class for an easy separation (see a comment thread in the review of this PR for a more detailed description). I can take another stab at this though if others feel strongly about it!

**Review by briandleahy — APPROVED** [review]
Looks good to me! I'm happy to merge it in if you're ready. 

With regards to a separate map entry class -- I think this is still a good goal, but we can wait to do it until later. The best is the enemy of the good :smile:

I'll merge this in at the end of the day today, unless you let me know you want me to wait.

**Comment by barkls** [comment]
Hold off on merging. I've discovered a bug with loading old Scatterers
objects since they no longer have ties.

**Comment by barkls** [comment]
Ok I fixed the backwards compatibility bug.

**Review by briandleahy — APPROVED** [review]

**Pull request merged**
